### PR TITLE
[shopsys] top products are now in elasticsearch

### PR DESCRIPTION
**Pull request by AndrejBlaho**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Top products data are now stored in and queried on frontend from elasticsearch which should lead to better performance.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://ab-top-products-in-elastic.odin.shopsys.cloud
  - https://cz.ab-top-products-in-elastic.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 1]
```diff
@@ -62,6 +65,8 @@ public function __construct(
         protected readonly ProductExportFieldProvider $productExportFieldProvider,
         protected readonly PricingGroupSettingFacade $pricingGroupSettingFacade,
         protected readonly ProductRepository $productRepository,
+        protected readonly TopProductFacade $topProductFacade,
+        protected array $domainTopProductsIndexedByProductId = [],
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: This should be only class property not dependency as it is not provided via dependency injection

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 2]
```diff
@@ -145,6 +150,8 @@ protected function extractResult(Product $product, int $domainId, string $locale
      */
     protected function getExportedFieldValue(int $domainId, Product $product, string $locale, string $field): mixed
     {
+        $topProduct = $this->domainTopProductsIndexedByProductId[$product->getId()] ?? null;
+
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: It would be better to have a method for this to check if top products are loaded and load them if not and then return the value.

File: project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php [hunk 3]
```diff
@@ -4,48 +4,14 @@
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php]: This class can be removed now, right?

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 4]
```diff
@@ -62,6 +65,8 @@ public function __construct(
         protected readonly ProductExportFieldProvider $productExportFieldProvider,
         protected readonly PricingGroupSettingFacade $pricingGroupSettingFacade,
         protected readonly ProductRepository $productRepository,
+        protected readonly TopProductFacade $topProductFacade,
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: This should be mentioned in upgrade notes

File: packages/frontend-api/src/Model/Resolver/Products/PromotedProductsQuery.php [hunk 5]
```diff
@@ -15,22 +16,21 @@ class PromotedProductsQuery extends AbstractQuery
      * @param \Shopsys\FrameworkBundle\Model\Product\TopProduct\TopProductFacade $topProductFacade
      * @param \Shopsys\FrameworkBundle\Component\Domain\Domain $domain
      * @param \Shopsys\FrameworkBundle\Model\Customer\User\CurrentCustomerUser $currentCustomerUser
+     * @param \Shopsys\FrameworkBundle\Model\Product\ProductElasticsearchProvider $productElasticsearchProvider
      */
     public function __construct(
         protected readonly TopProductFacade $topProductFacade,
         protected readonly Domain $domain,
         protected readonly CurrentCustomerUser $currentCustomerUser,
+        protected readonly ProductElasticsearchProvider $productElasticsearchProvider,
```
Comment (TomasLudvik) [inline packages/frontend-api/src/Model/Resolver/Products/PromotedProductsQuery.php]: This should be mentioned in the upgrade notes

**Review by TomasLudvik — CHANGES_REQUESTED** [review]

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php (hunk 1 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: what about $variantCache bellow? should that be changed as well?

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php (hunk 2 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: I rewrote it. Is this how you meant it?

File: project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php (hunk 3 above)
Comment (AndrejBlaho) [reply project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php]: I remember brief discussion with @vitek-rostislav about this and conclusion was keep it as it is for some reason, but it makes sense to remove it as you suggest, so i deleted it for now. But i am not sure about the reasoning and most correct approach so please clarify.

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php (hunk 4 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: added

File: packages/frontend-api/src/Model/Resolver/Products/PromotedProductsQuery.php (hunk 5 above)
Comment (AndrejBlaho) [reply packages/frontend-api/src/Model/Resolver/Products/PromotedProductsQuery.php]: added as well

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php (hunk 1 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: Yes, but you will have to mention it in UPGRADE notes.

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 6]
```diff
@@ -25,11 +25,15 @@
 use Shopsys\FrameworkBundle\Model\Product\ProductRepository;
 use Shopsys\FrameworkBundle\Model\Product\ProductVisibility;
 use Shopsys\FrameworkBundle\Model\Product\ProductVisibilityFacade;
+use Shopsys\FrameworkBundle\Model\Product\TopProduct\TopProductFacade;
 use Shopsys\FrameworkBundle\Model\Seo\HreflangLinksFacade;
 use Symfony\Contracts\Service\ResetInterface;
 
 class ProductExportRepository implements ResetInterface
 {
+    protected array $domainTopProductsIndexedByProductId = [];
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]:
```suggestion
    /**
      * @property array<int, \Shopsys\FrameworkBundle\Model\Product\TopProduct\TopProduct>|null
      */
    protected ?array $domainTopProductsIndexedByProductId = null;
```

Comment (TomasLudvik) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: I have wrote this suggestion by hand, you might need to correct it :)

File: project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php (hunk 3 above)
Comment (vitek-rostislav) [reply project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php]: well, the general reason is that when moving logic from project-base to the packages, it is easier to keep the empty class in project-base - we do not have a reverse function in the annotation fixer so we would have to rewrite all the usages manually then. Moreover, at least about some of the classes we can assume they will be overwritten on the particular project sooner or later (entities, data objects, facades, ...) so already extended empty project-base classes can help the developers a little in the future.

TLDR; I do not think the empty classes in the project-base can do any harm, rather it is the opposite.

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php (hunk 2 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: I would like you to use that method to also retrieve that product, so it could be named `getDomainTopProductById`-

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 7]
```diff
@@ -417,6 +433,20 @@ protected function getResults(QueryBuilder $queryBuilder, array $fields, int $do
         return $results;
     }
 
+    /**
+     * @param int $domainId
+     */
+    protected function loadTopProductsIfNotLoaded(int $domainId): void
+    {
+        if (empty($this->domainTopProductsIndexedByProductId)) {
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]: What if second domain ID is used?

**Review by TomasLudvik — CHANGES_REQUESTED** [review]

File: project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php (hunk 3 above)
Comment (TomasLudvik) [reply project-base/app/src/FrontendApi/Resolver/Products/PromotedProductsQuery.php]: My opinion is, that it is ok to have empty classes for the basic functionality, that is always changed (entity, factory,...) but I would remove it for other less common scenarios. Reasons:
- searching for class is duplicate and you need to check if the functionality is extended or that class is empty
- annotation fixer causes changes in files that do not need to be reviewed
- projects have empty classes that are actually not extended

File: packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php [hunk 8]
```diff
@@ -417,6 +432,24 @@ protected function getResults(QueryBuilder $queryBuilder, array $fields, int $do
...
     }
 
+    /**
+     * @param int $productId
+     * @param int $domainId
+     * @return \Shopsys\FrameworkBundle\Model\Product\TopProduct\TopProduct|null
+     */
+    protected function getTopProductByDomainId(int $productId, int $domainId): ?TopProduct
+    {
+        if (!isset($this->domainTopProductsIndexedByDomainIdAndProductId[$domainId])) {
```
Comment (TomasLudvik) [inline packages/framework/src/Model/Product/Elasticsearch/ProductExportRepository.php]:
```suggestion
        if (!array_key_exists($domainId, $this->domainTopProductsIndexedByDomainIdAndProductId)) {
```

**Review by TomasLudvik — APPROVED** [review]

**Review by TomasLudvik — APPROVED** [review]

**Review by pk16011990 — APPROVED** [review]
